src/main/scala/com.mediative.eigenflow/process/ProcessSupervisor.scala

```diff
@@ -0,0 +1,47 @@
+package com.mediative.eigenflow.process
+
+import java.util.Date
+
+import akka.actor.SupervisorStrategy.Stop
+import akka.actor.{ Actor, ActorRef, OneForOneStrategy, Props, SupervisorStrategy, Terminated }
+import com.mediative.eigenflow.StagedProcess
+import com.mediative.eigenflow.publisher.MessagingSystem
+
+import scala.util.{ Failure, Success, Try }
+import scala.concurrent.ExecutionContext.Implicits.global
+
+object ProcessSupervisor {
+  def props(process: StagedProcess, date: Option[Date])(implicit messagingSystem: MessagingSystem) = Props(new ProcessSupervisor(process, date))
+
+  case object Done
+  case object Failed
+}
+
+class ProcessSupervisor(val process: StagedProcess, val startDate: Option[Date])(implicit val messagingSystem: MessagingSystem) extends Actor {
```

Comment:
@WadeWaldron thanks for the feedback. Let's try to be a bit more specific.

The current implementation is very close to what you describe, unless I misunderstood your point.

1. "ProcessSupervisor" is the `top level actor`: see https://github.com/dmitri-carpov/eigenflow/blob/master/src/main/scala/com.mediative.eigenflow/EigenflowBootstrap.scala#L47 

2. "ProcessSupervisor" creates "ProcessManager" and defines a supervision strategy for it, which simply stops "ProcessManager" in case of errors. Also it listens for "ProcessManager"'s death and shutdowns the actor system first, then does `System.exit` (since I couldn't find any better way to exit with the correct exit code, which should be non-zero in case of failures). 
The system shut down is controlled by the "ProcessSupervisor" only, thus it is the only actor responsible for cleaning up and correct system exit.

see https://github.com/dmitri-carpov/eigenflow/blob/master/src/main/scala/com.mediative.eigenflow/process/ProcessSupervisor.scala#L39

That function is very close to what you called hook, unless you meant another technique.
btw: "ProcessSupervisor" is very close to what they call "Reaper" in http://letitcrash.com/post/30165507578/shutdown-patterns-in-akka-2

3. The current actor hierarchy is build on the following principle:

"ProcessSupervisor" (1 top level actor) - creates & monitors -> "ProcessManager" (1 persistence actor) - creates and monitors -> "ProcessFSM" (actor per processing date)

"ProcessManager" escalates all problems to "ProcessSupervisor"
"ProcessSupervisor" shuts down "ProcessManager" in case of problems
"ProcessSupervisor" shuts down the system either when "ProcessManager" is dead or when "ProcessManager" explicitly asks to shut down via sending "Done" or "Failed"


Also regarding brut force of system exit, "ProcessSupervisor" controls this process and as I mentioned above it tries to make it as clean as possible by shutting down actor system first and waiting until it's terminated before calling `System.exit`, would be nice to see a cleaner way.